pkg/virt-operator/resource/apply/delete.go

```diff
@@ -461,3 +450,151 @@ func DeleteAll(kv *v1.KubeVirt,
+			return true
+		}
+	}
+	return false
+}
+
+func crdFilterNeedFinalizerAdded(crds []*extv1.CustomResourceDefinition) []*extv1.CustomResourceDefinition {
+	filtered := []*extv1.CustomResourceDefinition{}
+
+	for _, crd := range crds {
+		if crd.DeletionTimestamp == nil && !controller.HasFinalizer(crd, v1.VirtOperatorComponentFinalizer) {
+			filtered = append(filtered, crd)
+		}
+	}
+
+	return filtered
+}
+
+func crdFilterNeedDeletion(crds []*extv1.CustomResourceDefinition) []*extv1.CustomResourceDefinition {
+	filtered := []*extv1.CustomResourceDefinition{}
+
+	for _, crd := range crds {
+		if crd.DeletionTimestamp == nil {
+			filtered = append(filtered, crd)
+		}
+	}
+	return filtered
+}
+
+func crdFilterNeedFinalizerRemoved(crds []*extv1.CustomResourceDefinition) []*extv1.CustomResourceDefinition {
+	filtered := []*extv1.CustomResourceDefinition{}
+	for _, crd := range crds {
+		if !crdInstanceDeletionCompleted(crd) {
+			// All crds must have all crs removed before any CRD finalizer can be removed
+			return []*extv1.CustomResourceDefinition{}
+		} else if controller.HasFinalizer(crd, v1.VirtOperatorComponentFinalizer) {
+			filtered = append(filtered, crd)
+		}
+	}
+	return filtered
+}
+
+func crdHandleDeletion(kvkey string,
+	stores util.Stores,
+	clientset kubecli.KubevirtClient,
+	expectations *util.Expectations) error {
+
+	ext := clientset.ExtensionsClient()
+	objects := stores.CrdCache.List()
+
+	finalizerPath := "/metadata/finalizers"
+
+	crds := []*extv1.CustomResourceDefinition{}
+	for _, obj := range objects {
+		crd, ok := obj.(*extv1.CustomResourceDefinition)
+		if !ok {
+			log.Log.Errorf(castFailedFmt, obj)
+			return nil
+		}
+		crds = append(crds, crd)
+	}
+
+	needFinalizerAdded := crdFilterNeedFinalizerAdded(crds)
+	needDeletion := crdFilterNeedDeletion(crds)
+	needFinalizerRemoved := crdFilterNeedFinalizerRemoved(crds)
+
+	for _, crd := range needFinalizerAdded {
+		crdCopy := crd.DeepCopy()
+		controller.AddFinalizer(crdCopy, v1.VirtOperatorComponentFinalizer)
+
+		patchBytes, err := json.Marshal(crdCopy.Finalizers)
+		if err != nil {
+			return err
+		}
+		ops := fmt.Sprintf(`[{ "op": "add", "path": "%s", "value": %s }]`, finalizerPath, string(patchBytes))
+		_, err = ext.ApiextensionsV1().CustomResourceDefinitions().Patch(context.Background(), crd.Name, types.JSONPatchType, []byte(ops), metav1.PatchOptions{})
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, crd := range needDeletion {
+		key, err := controller.KeyFunc(crd)
+		if err != nil {
+			return err
+		}
+
+		expectations.Crd.AddExpectedDeletion(kvkey, key)
+		err = ext.ApiextensionsV1().CustomResourceDefinitions().Delete(context.Background(), crd.Name, metav1.DeleteOptions{})
+		if err != nil {
+			expectations.Crd.DeletionObserved(kvkey, key)
+			log.Log.Errorf("Failed to delete crd %+v: %v", crd, err)
+			return err
+		}
+	}
+
+	for _, crd := range needFinalizerRemoved {
+		var ops string
+		if len(crd.Finalizers) > 1 {
```

Comment:
What happens if something else adds a finalizer? Would we then repeatedly try to patch the object? How about using `if controller.hasFinalizer()`?

Reply:
its not obvious in this line, but the "needsFinalizerRemoved" list will only add a crd if the crd has the exact finalizer we're removing.